x-pack/filebeat/input/awss3/sqs.go

```diff
@@ -76,37 +82,77 @@ func (r *sqsReader) Receive(ctx context.Context) error {
 		// Release unused workers.
 		r.workerSem.Release(workers - len(msgs))
 
-		// Process each SQS message asynchronously with a goroutine.
 		r.log.Debugf("Received %v SQS messages.", len(msgs))
 		r.metrics.sqsMessagesReceivedTotal.Add(uint64(len(msgs)))
-		workerWg.Add(len(msgs))
+
+		// Add to processing wait group to wait for all messages to be processed.
+		processingWg.Add(len(msgs))
+		deletionWg.Add(len(msgs))
 
 		for _, msg := range msgs {
+			// Process each SQS message asynchronously with a goroutine.
 			go func(msg types.Message, start time.Time) {
 				id := r.metrics.beginSQSWorker()
 				defer func() {
+					// Mark processing wait group as done.
 					r.metrics.endSQSWorker(id)
-					workerWg.Done()
+					processingWg.Done()
 					r.workerSem.Release(1)
 				}()
 
-				if err := r.msgHandler.ProcessSQS(ctx, &msg); err != nil {
+				// Create a pipeline client scoped to this goroutine.
+				client, err := r.pipeline.ConnectWith(beat.ClientConfig{
```

Comment:
Rather than creating a new goroutine for each message, are you able to reuse previously created goroutines? This would require being able to reset the ack handler at least.

This should generate considerably less garbage and minimize repeated work each time you process a message.

Each worker could receive the message to process on a channel. 

You could perhaps use a buffered channel of available worker channels for distributing messages to channels. The buffer size would be equal to `maxMessagesInflight` and the it would initially be populated with the receive channel for each worker. When a worker is in use it received from this channel, when it is free again it sends to that channel.

Instead of using a semaphore you'd just be receiving from the channel of worker channels.

Reply:
> You could perhaps use a buffered channel of available worker channels for distributing messages to workers. The buffer size would be equal to `maxMessagesInflight` and it would initially be populated with the receive channel for each worker. When a worker is in use it received from this channel, when it is free again it sends to that channel.

Got it, let me implement it and compare memory usage.

As for context, let me sum up the benchmark comparison I run between 18.2.2 and the current refactoring rebased on 18.12.2.
The type load was 1 SQS message : 1 S3 object : 1 event, for a total of 81349 event.
- 18.2.2:  2h 20 mins taken ca for full ingestion, 82MB RAM, 10 EPS avg ca, 100 messages inflight constant
- current refactoring: 8 mins taken ca for full ingestion, 645 MB RAM, 320 EPS avg ca, thousands of messages inflight varying

So yes, there's a 8x memory usage. And we should aim to generate less garbage.
More specifically, there must be some leaks around I was not able to identify yet, because memory usage does not reduce after the full ingestion.

Could you help me to identify the leak?

Reply:
> Got it, let me implement it and compare memory usage.

It was quick to implement, but unless I didn't get it right this just caps again a constant number of messages inflight

```go
	// Start a fixed amount of goroutines that will process all the SQS messages sent to the processingChan asynchronously.
	processingChan := make(chan processingData, r.maxMessagesInflight)

	for i := 0; i < r.maxMessagesInflight; i++ {
		go func() {
			incomingData := <-processingChan
			msg := incomingData.msg
			start := incomingData.start

			[same code as before]
		}()
	}


	for ctx.Err() == nil {
		// Receive (at most) as many SQS messages as there are workers.
		msgs, err := r.sqs.ReceiveMessage(ctx, r.maxMessagesInflight)
		if err != nil {
			if ctx.Err() == nil {
				r.log.Warnw("SQS ReceiveMessage returned an error. Will retry after a short delay.", "error", err)

				// Throttle retries.
				_ = timed.Wait(ctx, sqsRetryDelay)
			}
			continue
		}
		
		r.log.Debugf("Received %v SQS messages.", len(msgs))
		r.metrics.sqsMessagesReceivedTotal.Add(uint64(len(msgs)))

		// Add to processing wait group to wait for all messages to be processed.
		processingWg.Add(len(msgs))
		deletionWg.Add(len(msgs))

		for _, msg := range msgs {
			// Send each SQS message to a channel where a fixed amount of goroutines will process all of them asynchronously.
			processingChan <- processingData{msg: &msg, start: time.Now()}
		}
	}
```
, but unless I missed something

Reply:
> Could you help me to identify the leak?

We need heap profiles while the benchmarks are running. The best place to get those is the `/debug/pprof` profiling HTTP endpoint. We'd probably want CPU profiles as well to give us hints about what to improve.

Reply:
Also, the change you have made is naturally going to use more memory because you are able to fill the entire Beat memory queue. You are going from 100 buffered events to possibly the 4096 queue size limit. 

This could explain the memory usage increase, but we'd need profiles to confirm.

Reply:
@cmacknz , yes, that memory usage I expect to mainly increase because of filling the queue size limit.
I was looking into the reason why memory does not goes down, and I suspect is because the ackers are not garbage collected. But I cannot see that from the heap profile I've already checked.
I will try to collect another one and share

Reply:
```
(pprof) top                                                                     
Showing nodes accounting for 19766.83kB, 66.92% of 29538.93kB total
Showing top 10 nodes out of 187
      flat  flat%   sum%        cum   cum%
 5456.13kB 18.47% 18.47%  5456.13kB 18.47%  github.com/goccy/go-json/internal/decoder.init.0
 3786.34kB 12.82% 31.29%  3786.34kB 12.82%  github.com/elastic/beats/v7/libbeat/asset.GetFields
 2741.30kB  9.28% 40.57%  2741.30kB  9.28%  github.com/goccy/go-json/internal/encoder.init.0
 2061.41kB  6.98% 47.55%  3597.58kB 12.18%  github.com/elastic/beats/v7/x-pack/filebeat/input/awss3.(*s3ObjectProcessor).readJSON
 1536.62kB  5.20% 52.75%  1536.62kB  5.20%  github.com/aws/aws-sdk-go-v2/service/sqs.awsAwsquery_deserializeDocumentMessage
 1024.72kB  3.47% 56.22%  1024.72kB  3.47%  google.golang.org/protobuf/internal/filedesc.(*Enum).unmarshalSeed
 1024.09kB  3.47% 59.69%  1024.09kB  3.47%  github.com/armon/go-radix.(*node).addEdge
  902.59kB  3.06% 62.74%  2136.21kB  7.23%  compress/flate.NewWriter
  650.62kB  2.20% 64.94%  1233.63kB  4.18%  compress/flate.(*compressor).init
  583.01kB  1.97% 66.92%   583.01kB  1.97%  compress/flate.newDeflateFast (inline)
(pprof) list readJSON                                                           
Total: 28.85MB    
ROUTINE ======================== github.com/elastic/beats/v7/x-pack/filebeat/input/awss3.(*s3ObjectProcessor).readJSON in /Users/andreaspacca/Sources/beats/x-pack/filebeat/input/awss3/s3_objects.go
    2.01MB     3.51MB (flat, cum) 12.18% of Total
         .          .    236:func (p *s3ObjectProcessor) readJSON(r io.Reader) error {
         .          .    237:   dec := json.NewDecoder(r)
         .          .    238:   dec.UseNumber()
         .          .    239:
         .          .    240:   for dec.More() && p.ctx.Err() == nil {
         .          .    241:           offset := dec.InputOffset()
         .          .    242:
         .          .    243:           var item json.RawMessage
         .          .    244:           if err := dec.Decode(&item); err != nil {
         .          .    245:                   return fmt.Errorf("failed to decode json: %w", err)
         .          .    246:           }
         .          .    247:
         .          .    248:           if p.readerConfig.ExpandEventListFromField != "" {
         .          .    249:                   if err := p.splitEventList(p.readerConfig.ExpandEventListFromField, item, offset, p.s3ObjHash); err != nil {
         .          .    250:                           return err
         .          .    251:                   }
         .          .    252:                   continue
         .          .    253:           }
         .          .    254:
         .          .    255:           data, _ := item.MarshalJSON()
    2.01MB     2.01MB    256:           evt := p.createEvent(string(data), offset)
         .     1.50MB    257:           p.publish(p.acker, &evt)
         .          .    258:   }
         .          .    259:
         .          .    260:   return nil
         .          .    261:}
         .          .    262:
(pprof) list publish          
Total: 28.85MB
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/outputs/elasticsearch.(*Client).publishEvents in /Users/andreaspacca/Sources/beats/libbeat/outputs/elasticsearch/client.go
         0     2.09MB (flat, cum)  7.23% of Total
         .          .    221:func (client *Client) publishEvents(ctx context.Context, data []publisher.Event) ([]publisher.Event, error) {
         .          .    222:   span, ctx := apm.StartSpan(ctx, "publishEvents", "output")
         .          .    223:   defer span.End()
         .          .    224:   begin := time.Now()
         .          .    225:   st := client.observer
         .          .    226:
         .          .    227:   if st != nil {
         .          .    228:           st.NewBatch(len(data))
         .          .    229:   }
         .          .    230:
         .          .    231:   if len(data) == 0 {
         .          .    232:           return nil, nil
         .          .    233:   }
         .          .    234:
         .          .    235:   // encode events into bulk request buffer, dropping failed elements from
         .          .    236:   // events slice
         .          .    237:   origCount := len(data)
         .          .    238:   span.Context.SetLabel("events_original", origCount)
         .          .    239:   data, bulkItems := client.bulkEncodePublishRequest(client.conn.GetVersion(), data)
         .          .    240:   newCount := len(data)
         .          .    241:   span.Context.SetLabel("events_encoded", newCount)
         .          .    242:   if st != nil && origCount > newCount {
         .          .    243:           st.Dropped(origCount - newCount)
         .          .    244:   }
         .          .    245:   if newCount == 0 {
         .          .    246:           return nil, nil
         .          .    247:   }
         .          .    248:
         .          .    249:   params := map[string]string{"filter_path": "errors,items.*.error,items.*.status"}
         .     2.09MB    250:   status, result, sendErr := client.conn.Bulk(ctx, "", "", params, bulkItems)
         .          .    251:
         .          .    252:   if sendErr != nil {
         .          .    253:           if status == http.StatusRequestEntityTooLarge {
         .          .    254:                   // This error must be handled by splitting the batch, propagate it
         .          .    255:                   // back to Publish instead of reporting it directly
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/pipeline.(*client).Publish in /Users/andreaspacca/Sources/beats/libbeat/publisher/pipeline/client.go
         0     1.50MB (flat, cum)  5.20% of Total
         .          .     73:func (c *client) Publish(e beat.Event) {
         .          .     74:   c.mutex.Lock()
         .          .     75:   defer c.mutex.Unlock()
         .          .     76:
         .     1.50MB     77:   c.publish(e)
         .          .     78:}
         .          .     79:
         .          .     80:func (c *client) publish(e beat.Event) {
         .          .     81:   var (
         .          .     82:           event   = &e
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/pipeline.(*client).publish in /Users/andreaspacca/Sources/beats/libbeat/publisher/pipeline/client.go
         0     1.50MB (flat, cum)  5.20% of Total
         .          .     80:func (c *client) publish(e beat.Event) {
         .          .     81:   var (
         .          .     82:           event   = &e
         .          .     83:           publish = true
         .          .     84:   )
         .          .     85:
         .          .     86:   c.onNewEvent()
         .          .     87:
         .          .     88:   if !c.isOpen.Load() {
         .          .     89:           // client is closing down -> report event as dropped and return
         .          .     90:           c.onDroppedOnPublish(e)
         .          .     91:           return
         .          .     92:   }
         .          .     93:
         .          .     94:   if c.processors != nil {
         .          .     95:           var err error
         .          .     96:
         .     1.50MB     97:           event, err = c.processors.Run(event)
         .          .     98:           publish = event != nil
         .          .     99:           if err != nil {
         .          .    100:                   // If we introduce a dead-letter queue, this is where we should
         .          .    101:                   // route the event to it.
         .          .    102:                   c.logger.Errorf("Failed to publish event: %v", err)
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/pipeline.(*netClientWorker).publishBatch in /Users/andreaspacca/Sources/beats/libbeat/publisher/pipeline/client_worker.go
         0     2.09MB (flat, cum)  7.23% of Total
         .          .    162:func (w *netClientWorker) publishBatch(batch publisher.Batch) error {
         .          .    163:   ctx := context.Background()
         .          .    164:   if w.tracer != nil && w.tracer.Recording() {
         .          .    165:           tx := w.tracer.StartTransaction("publish", "output")
         .          .    166:           defer tx.End()
         .          .    167:           tx.Context.SetLabel("worker", "netclient")
         .          .    168:           ctx = apm.ContextWithTransaction(ctx, tx)
         .          .    169:   }
         .     2.09MB    170:   err := w.client.Publish(ctx, batch)
         .          .    171:   if err != nil {
         .          .    172:           err = fmt.Errorf("failed to publish events: %w", err)
         .          .    173:           apm.CaptureError(ctx, err).Send()
         .          .    174:           w.logger.Error(err)
         .          .    175:           // on error return to connect loop
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/pipeline.(*netClientWorker).run in /Users/andreaspacca/Sources/beats/libbeat/publisher/pipeline/client_worker.go
         0     2.09MB (flat, cum)  7.23% of Total
         .          .    112:func (w *netClientWorker) run() {
         .          .    113:   var (
         .          .    114:           connected         = false
         .          .    115:           reconnectAttempts = 0
         .          .    116:   )
         .          .    117:
         .          .    118:   for {
         .          .    119:           // We wait for either the worker to be closed or for there to be a batch of
         .          .    120:           // events to publish.
         .          .    121:           select {
         .          .    122:
         .          .    123:           case <-w.done:
         .          .    124:                   return
         .          .    125:
         .          .    126:           case batch := <-w.qu:
         .          .    127:                   if batch == nil {
         .          .    128:                           continue
         .          .    129:                   }
         .          .    130:
         .          .    131:                   // Try to (re)connect so we can publish batch
         .          .    132:                   if !connected {
         .          .    133:                           // Return batch to other output workers while we try to (re)connect
         .          .    134:                           batch.Cancelled()
         .          .    135:
         .          .    136:                           if reconnectAttempts == 0 {
         .          .    137:                                   w.logger.Infof("Connecting to %v", w.client)
         .          .    138:                           } else {
         .          .    139:                                   w.logger.Infof("Attempting to reconnect to %v with %d reconnect attempt(s)", w.client, reconnectAttempts)
         .          .    140:                           }
         .          .    141:
         .          .    142:                           err := w.client.Connect()
         .          .    143:                           connected = err == nil
         .          .    144:                           if connected {
         .          .    145:                                   w.logger.Infof("Connection to %v established", w.client)
         .          .    146:                                   reconnectAttempts = 0
         .          .    147:                           } else {
         .          .    148:                                   w.logger.Errorf("Failed to connect to %v: %v", w.client, err)
         .          .    149:                                   reconnectAttempts++
         .          .    150:                           }
         .          .    151:
         .          .    152:                           continue
         .          .    153:                   }
         .          .    154:
         .     2.09MB    155:                   if err := w.publishBatch(batch); err != nil {
         .          .    156:                           connected = false
         .          .    157:                   }
         .          .    158:           }
         .          .    159:   }
         .          .    160:}
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/pipetool.(*connectEditPipeline).ConnectWith in /Users/andreaspacca/Sources/beats/libbeat/publisher/pipetool/pipetool.go
         0        1MB (flat, cum)  3.47% of Total
         .          .     40:func (p *connectEditPipeline) ConnectWith(cfg beat.ClientConfig) (beat.Client, error) {
         .        1MB     41:   cfg, err := p.edit(cfg)
         .          .     42:   if err != nil {
         .          .     43:           return nil, err
         .          .     44:   }
         .          .     45:   return p.parent.ConnectWith(cfg)
         .          .     46:}
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/processing.(*group).Run in /Users/andreaspacca/Sources/beats/libbeat/publisher/processing/processors.go
         0     1.50MB (flat, cum)  5.20% of Total
         .          .    121:func (p *group) Run(event *beat.Event) (*beat.Event, error) {
         .          .    122:   if p == nil || len(p.list) == 0 {
         .          .    123:           return event, nil
         .          .    124:   }
         .          .    125:
         .          .    126:   for _, sub := range p.list {
         .          .    127:           var err error
         .          .    128:
         .     1.50MB    129:           event, err = sub.Run(event)
         .          .    130:           if err != nil {
         .          .    131:                   // XXX: We don't drop the event, but continue filtering here if the most
         .          .    132:                   //      recent processor did return an event.
         .          .    133:                   //      We want processors having this kind of implicit behavior
         .          .    134:                   //      on errors?
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/processing.(*processorFn).Run in /Users/andreaspacca/Sources/beats/libbeat/publisher/processing/processors.go
         0   512.14kB (flat, cum)  1.73% of Total
         .   512.14kB    159:func (p *processorFn) Run(e *beat.Event) (*beat.Event, error) { return p.fn(e) }
         .          .    160:
         .          .    161:func clientEventMeta(meta mapstr.M, needsCopy bool) *processorFn {
         .          .    162:   fn := func(event *beat.Event) { addMeta(event, meta) }
         .          .    163:   if needsCopy {
         .          .    164:           fn = func(event *beat.Event) { addMeta(event, meta.Clone()) }
ROUTINE ======================== github.com/elastic/beats/v7/x-pack/filebeat/input/awss3.(*s3ObjectProcessor).publish in /Users/andreaspacca/Sources/beats/x-pack/filebeat/input/awss3/s3_objects.go
         0     1.50MB (flat, cum)  5.20% of Total
         .          .    392:func (p *s3ObjectProcessor) publish(ack *EventACKTracker, event *beat.Event) {
         .          .    393:   event.Private = ack
         .          .    394:   p.s3EventsCreatedTotal++
         .          .    395:   p.metrics.s3EventsCreatedTotal.Inc()
         .     1.50MB    396:   p.publisher.Publish(*event)
         .          .    397:}
         .          .    398:
         .          .    399:func (p *s3ObjectProcessor) createEvent(message string, offset int64) beat.Event {
         .          .    400:   event := beat.Event{
         .          .    401:           Timestamp: time.Now().UTC(),
(pprof) list process.+Run  
Total: 28.85MB
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/processors.(*SafeProcessor).Run in /Users/andreaspacca/Sources/beats/libbeat/processors/safe_processor.go
         0   512.14kB (flat, cum)  1.73% of Total
         .          .     37:func (p *SafeProcessor) Run(event *beat.Event) (*beat.Event, error) {
         .          .     38:   if atomic.LoadUint32(&p.closed) == 1 {
         .          .     39:           return nil, ErrClosed
         .          .     40:   }
         .   512.14kB     41:   return p.Processor.Run(event)
         .          .     42:}
         .          .     43:
         .          .     44:// Close makes sure the underlying `Close` function is called only once.
         .          .     45:func (p *SafeProcessor) Close() (err error) {
         .          .     46:   if atomic.CompareAndSwapUint32(&p.closed, 0, 1) {
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/processors/actions.(*addFields).Run in /Users/andreaspacca/Sources/beats/libbeat/processors/actions/add_fields.go
         0   512.02kB (flat, cum)  1.73% of Total
         .          .     75:func (af *addFields) Run(event *beat.Event) (*beat.Event, error) {
         .          .     76:   if event == nil || len(af.fields) == 0 {
         .          .     77:           return event, nil
         .          .     78:   }
         .          .     79:
         .          .     80:   fields := af.fields
         .          .     81:   if af.shared {
         .   512.02kB     82:           fields = fields.Clone()
         .          .     83:   }
         .          .     84:
         .          .     85:   if af.overwrite {
         .          .     86:           event.DeepUpdate(fields)
         .          .     87:   } else {
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/processors/actions.(*extract_field).Run in /Users/andreaspacca/Sources/beats/libbeat/processors/actions/extract_field.go
         0   512.14kB (flat, cum)  1.73% of Total
         .          .     75:func (f *extract_field) Run(event *beat.Event) (*beat.Event, error) {
         .          .     76:   fieldValue, err := event.GetValue(f.Field)
         .          .     77:   if err != nil {
         .          .     78:           return event, fmt.Errorf("error getting field '%s' from event", f.Field)
         .          .     79:   }
         .          .     80:
         .          .     81:   value, ok := fieldValue.(string)
         .          .     82:   if !ok {
         .          .     83:           return event, fmt.Errorf("could not get a string from field '%s'", f.Field)
         .          .     84:   }
         .          .     85:
         .          .     86:   parts := strings.Split(value, f.Separator)
         .          .     87:   parts = deleteEmpty(parts)
         .          .     88:   if len(parts) < f.Index+1 {
         .          .     89:           return event, fmt.Errorf("index is out of range for field '%s'", f.Field)
         .          .     90:   }
         .          .     91:
         .   512.14kB     92:   _, _ = event.PutValue(f.Target, parts[f.Index])
         .          .     93:
         .          .     94:   return event, nil
         .          .     95:}
         .          .     96:
         .          .     97:func (f extract_field) String() string {
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/processors/add_docker_metadata.(*addDockerMetadata).Run in /Users/andreaspacca/Sources/beats/libbeat/processors/add_docker_metadata/add_docker_metadata.go
         0   512.14kB (flat, cum)  1.73% of Total
         .          .    136:func (d *addDockerMetadata) Run(event *beat.Event) (*beat.Event, error) {
         .          .    137:   if !d.dockerAvailable {
         .          .    138:           return event, nil
         .          .    139:   }
         .          .    140:   var cid string
         .          .    141:   var err error
         .          .    142:
         .          .    143:   // Extract CID from the filepath contained in the "log.file.path" field.
         .          .    144:   if d.sourceProcessor != nil {
         .          .    145:           lfp, _ := event.Fields.GetValue("log.file.path")
         .          .    146:           if lfp != nil {
         .   512.14kB    147:                   event, err = d.sourceProcessor.Run(event)
         .          .    148:                   if err != nil {
         .          .    149:                           d.log.Debugf("Error while extracting container ID from source path: %v", err)
         .          .    150:                           return event, nil
         .          .    151:                   }
         .          .    152:
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/processors/add_formatted_index.(*AddFormattedIndex).Run in /Users/andreaspacca/Sources/beats/libbeat/processors/add_formatted_index/add_formatted_index.go
  512.01kB   512.01kB (flat, cum)  1.73% of Total
         .          .     60:func (p *AddFormattedIndex) Run(event *beat.Event) (*beat.Event, error) {
         .          .     61:   var index string
         .          .     62:   var err error
         .          .     63:   if p.fullEvent {
         .          .     64:           index, err = p.formatString.RunEvent(event)
         .          .     65:   } else {
         .          .     66:           index, err = p.formatString.Run(event.Timestamp)
         .          .     67:   }
         .          .     68:   if err != nil {
         .          .     69:           return nil, err
         .          .     70:   }
         .          .     71:
         .          .     72:   if event.Meta == nil {
         .          .     73:           event.Meta = mapstr.M{}
         .          .     74:   }
  512.01kB   512.01kB     75:   event.Meta[events.FieldMetaRawIndex] = index
         .          .     76:   return event, nil
         .          .     77:}
         .          .     78:
         .          .     79:func (p *AddFormattedIndex) String() string {
         .          .     80:   return fmt.Sprintf("add_index_pattern=%v", p.formatString)
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/processing.(*group).Run in /Users/andreaspacca/Sources/beats/libbeat/publisher/processing/processors.go
         0     1.50MB (flat, cum)  5.20% of Total
         .          .    121:func (p *group) Run(event *beat.Event) (*beat.Event, error) {
         .          .    122:   if p == nil || len(p.list) == 0 {
         .          .    123:           return event, nil
         .          .    124:   }
         .          .    125:
         .          .    126:   for _, sub := range p.list {
         .          .    127:           var err error
         .          .    128:
         .     1.50MB    129:           event, err = sub.Run(event)
         .          .    130:           if err != nil {
         .          .    131:                   // XXX: We don't drop the event, but continue filtering here if the most
         .          .    132:                   //      recent processor did return an event.
         .          .    133:                   //      We want processors having this kind of implicit behavior
         .          .    134:                   //      on errors?
ROUTINE ======================== github.com/elastic/beats/v7/libbeat/publisher/processing.(*processorFn).Run in /Users/andreaspacca/Sources/beats/libbeat/publisher/processing/processors.go
         0   512.14kB (flat, cum)  1.73% of Total
         .   512.14kB    159:func (p *processorFn) Run(e *beat.Event) (*beat.Event, error) { return p.fn(e) }
         .          .    160:
         .          .    161:func clientEventMeta(meta mapstr.M, needsCopy bool) *processorFn {
         .          .    162:   fn := func(event *beat.Event) { addMeta(event, meta) }
         .          .    163:   if needsCopy {
         .          .    164:           fn = func(event *beat.Event) { addMeta(event, meta.Clone()) }
(pprof) 

```

Reply:
I cannot see anything specific coming from the new code

Reply:
![profile001](https://github.com/elastic/beats/assets/221530/e6c89466-ae27-46e3-9cd0-365999f1722b)

Reply:
For these profiles, I notice the reported total in-use memory is only 30MB, which sounds reasonable, and much lower than the 4-600MB you mentioned. Are you still seeing larger amounts of memory use even after publication and acknowledgment is complete? If you provide the raw pprof files for the heap and allocs in that state, I could take a look.